File: lib/array-stack/array_stack.rb
class ArrayStack
  def initialize
    @array = []
    @index = 0
  end

  def empty?
    index.zero?
  end

  def push(val)
    array[index] = val
    @index += 1

    self
  end

  def pop
    if empty?
      nil
    else
      @index -= 1
      array[index]
    end
  end

  def inspect
    "/#{array.reverse.join(" ")}/"
  end

  private

  attr_reader :array, :index
end

File: lib/array-stack/array_stack.js
(function () {
  'use strict';

  var ArrayStack = function () {
    this.index = 0;
    this.array = [];
  };

  ArrayStack.prototype.isEmpty = function () {
    return !this.index;
  };

  ArrayStack.prototype.push = function (value) {
    this.array[this.index++] = value;

    return this;
  };

  ArrayStack.prototype.pop = function () {
    if (this.isEmpty()) return null;

    return this.array[--this.index];
  };

  module.exports = ArrayStack;
})();
